### PR TITLE
feat(vm): apply new controller design

### PR DESCRIPTION
**Pull request by yaroslavborbat**
## Description
Apply new VM controller design

## Checklist
- [ ] The code is covered by unit tests.
- [ ] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [X] Changes were tested in the Kubernetes cluster manually.

### CODE REVIEW

File: images/virtualization-artifact/pkg/controller/vm/internal/cpu.go [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+func (h *CPUHandler) setFinalizers(ctx context.Context, cpuModel *virtv2.VirtualMachineCPUModel) error {
+	return h.updateFinalizers(ctx, cpuModel, controllerutil.AddFinalizer)
+}
+
+func (h *CPUHandler) removeFinalizers(ctx context.Context, cpuModel *virtv2.VirtualMachineCPUModel) error {
+	return h.updateFinalizers(ctx, cpuModel, controllerutil.RemoveFinalizer)
+}
+
+func (h *CPUHandler) updateFinalizers(ctx context.Context, cpuModel *virtv2.VirtualMachineCPUModel, fn updaterFinalizer) error {
+	if cpuModel != nil && fn(cpuModel, virtv2.FinalizerVMCPUProtection) {
```
Comment (Isteb4k) [inline images/virtualization-artifact/pkg/controller/vm/internal/cpu.go:113]: How about using a ProtectionService to protect resources?

File: images/virtualization-artifact/pkg/controller/vm/internal/ipam.go [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+
+	// 2. Claim not found: create if possible or wait for the claim.
+	if ipAddressClaim == nil {
+		cb.Reason2(vmcondition.ReasonIPAddressClaimNotReady)
+		if current.Spec.VirtualMachineIPAddressClaim != "" {
+			h.logger.Info(fmt.Sprintf("The requested ip address claim (%s) for the virtual machine not found: waiting for the Claim", current.Spec.VirtualMachineIPAddressClaim))
+			cb.Message(fmt.Sprintf("The requested ip address claim (%s) for the virtual machine not found: waiting for the Claim", current.Spec.VirtualMachineIPAddressClaim))
+			return reconcile.Result{RequeueAfter: 2 * time.Second}, nil
+		}
+		//
```
Comment (Isteb4k) [inline images/virtualization-artifact/pkg/controller/vm/internal/ipam.go:143]: Empty comment

File: images/virtualization-artifact/pkg/controller/vm/internal/block_device.go [hunk 3]
```diff
@@ -0,0 +1,374 @@
...
+	}
+
+	if len(disks) == 0 {
+		return ""
+	}
+	return fmt.Sprintf("spec.blockDeviceRefs contain hotplugged disks: %s. Unplug or remove them from spec to continue.", strings.Join(disks, ", "))
+}
+
+// BlockDevicesReady check if all attached images and disks are ready to use by the VM.
+func (h *BlockDeviceHandler) BlockDevicesReady(vm *virtv2.VirtualMachine, s BlockDevicesState) (bool, int) {
```
Comment (Isteb4k) [inline images/virtualization-artifact/pkg/controller/vm/internal/block_device.go:177]: Not a public method
Mb rename to сountReadyBlockDevices?

File: images/virtualization-artifact/pkg/controller/vm/internal/block_device.go [hunk 4]
```diff
@@ -0,0 +1,374 @@
...
+	return h.updateFinalizers(ctx, vm, s, controllerutil.AddFinalizer)
+}
+
+// removeFinalizersOnBlockDevices remove protection finalizers on CVI,VI and VMD attached to the VM.
+func (h *BlockDeviceHandler) removeFinalizersOnBlockDevices(ctx context.Context, vm *virtv2.VirtualMachine, s BlockDevicesState) error {
+	return h.updateFinalizers(ctx, vm, s, controllerutil.RemoveFinalizer)
+}
+
+// updateFinalizers remove protection finalizers on CVI,VI and VMD attached to the VM.
+func (h *BlockDeviceHandler) updateFinalizers(ctx context.Context, vm *virtv2.VirtualMachine, s BlockDevicesState, fn updaterFinalizer) error {
```
Comment (Isteb4k) [inline images/virtualization-artifact/pkg/controller/vm/internal/block_device.go:219]: rename fn variable

File: images/virtualization-artifact/pkg/controller/vm/internal/lifecycle.go [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+	virtv2.MachineStopped,
+	virtv2.MachineTerminating,
+}
+
+type PhaseTransitions struct {
+	Slice []virtv2.VirtualMachinePhaseTransitionTimestamp
+	Map   map[virtv2.MachinePhase]virtv2.VirtualMachinePhaseTransitionTimestamp
+}
+
+func NewPhaseTransitions(phaseTransitions []virtv2.VirtualMachinePhaseTransitionTimestamp, oldPhase, newPhase virtv2.MachinePhase) PhaseTransitions {
```
Comment (Isteb4k) [inline images/virtualization-artifact/pkg/controller/vm/internal/lifecycle.go:299]: It's hard to understand what's going on here and why it's needed. Let's add a comment?

File: images/virtualization-artifact/pkg/controller/vm/internal/deletion_handler.go [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+	"github.com/deckhouse/virtualization-controller/pkg/controller/vm/internal/state"
+	"github.com/deckhouse/virtualization-controller/pkg/sdk/framework/helper"
+	virtv2 "github.com/deckhouse/virtualization/api/core/v1alpha2"
+)
+
+const nameDeletionHandler = "DeletionHandler"
+
+func NewDeletionHandler(client client.Client, logger *slog.Logger) *DeletionHandler {
+	if logger == nil {
```
Comment (Isteb4k) [inline images/virtualization-artifact/pkg/controller/vm/internal/deletion_handler.go:38]: Why do we check the logger for nil but not the client? It seems redundant to check for nil and reassign the logger here. Whichever one is passed, we should use.

**Review by Isteb4k — CHANGES_REQUESTED** [review]

File: images/virtualization-artifact/pkg/controller/vm/internal/deletion_handler.go [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+	return &DeletionHandler{client: client, logger: logger}
+}
+
+type DeletionHandler struct {
+	client client.Client
+	logger *slog.Logger
+}
+
+func (h *DeletionHandler) Handle(ctx context.Context, s state.VirtualMachineState) (reconcile.Result, error) {
+	if s.VirtualMachine().IsEmpty() || s.VirtualMachine().Current().GetDeletionTimestamp().IsZero() {
```
[the remark this answers is not in the document]
Comment (yaroslavborbat) [reply images/virtualization-artifact/pkg/controller/vm/internal/deletion_handler.go:50]: no, we return now if DeletionTimestamp is zero. In the isDeletion func we check the opposite

**Review by Isteb4k — APPROVED** [review]

**Pull request merged**